### PR TITLE
Equalization operator

### PR DESCRIPTION
**Pull request by stiepan**
Signed-off-by: Kamil Tokarski <ktokarski@nvidia.com>

## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
Adds per-channel histogram equalization operator (GPU, uint8_t). 

## Additional information:
Uses kernels added in the https://github.com/NVIDIA/DALI/pull/4565

It is needed to implement the standard RandAugment/AutoAugment policies.

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [x] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3187

### CODE REVIEW

File: dali/operators/image/color/equalize.cu [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      TensorListShape<2> ret{batch_size};
+      for (int sample_idx = 0; sample_idx < batch_size; sample_idx++) {
+        ret.set_tensor_shape(sample_idx, TensorShape<2>(shape[sample_idx].num_elements(), 1));
+      }
+      return ret;
+    }
+  }
+
+  kernels::KernelManager kmgr_;
+  kernels::KernelContext ctx_;
```
Comment (mzient) [inline dali/operators/image/color/equalize.cu:73]: KernelContext is an ephemeral object - in general, it's not stored for later.

**Review by mzient — APPROVED** [review]

File: dali/operators/image/color/equalize.cu [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    auto in_shape = GetFlattenedShape(in_view.shape);
+    TensorListView<StorageGPU, uint8_t, 2> out_view_flat{out_view.data, out_shape};
+    TensorListView<StorageGPU, const uint8_t, 2> in_view_flat{in_view.data, in_shape};
+    kmgr_.Run<Kernel>(0, ctx_, out_view_flat, in_view_flat);
+  }
+
+  template <int ndim>
+  TensorListShape<2> GetFlattenedShape(TensorListShape<ndim> shape) {
+    if (shape.sample_dim() == 3) {  // has_channels
+      return collapse_dims<2>(shape, {{0, shape.sample_dim() - 1}});
```
Comment (klecki) [inline dali/operators/image/color/equalize.cu:61]: The layout can be HWC and CHW, doesn't it impact anything here?

Comment (stiepan) [reply dali/operators/image/color/equalize.cu:61]: SequenceOperator handles the latter

**Review by klecki — APPROVED** [review]

File: dali/test/python/operator_2/test_equalize.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    pipe = images_pipeline(num_threads=4, device_id=0, batch_size=batch_size, layout=layout)
+    pipe.build()
+
+    for _ in range(num_iters):
+        equalized, imgs = pipe.run()
+        equalized = [np.array(img) for img in equalized.as_cpu()]
+        imgs = [np.array(img) for img in imgs.as_cpu()]
+        assert len(equalized) == len(imgs)
+        baseline = [equalize_cv_baseline(img, layout) for img in imgs]
+        check_batch(equalized, baseline)
```
[the remark this answers is not in the document]
Comment (stiepan) [reply dali/test/python/operator_2/test_equalize.py:69]: As discussed elsewhere, the algorithm is the same, the only float arithmetic is scaling of the comulative. It happens to work, let us see if the CI does not complain.

File: dali/operators/image/color/equalize.cu [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    auto input_type = input.type();
+    auto layout = input.GetLayout();
+    DALI_ENFORCE(input_type == type2id<uint8_t>::value,
+                 make_string("Unsupported input type for equalize operator: ", input_type,
+                             ". Expected input type: `uint8_t`."));
+    // enforced by the layouts specified in operator schema
+    assert(layout.size() == 2 || layout.size() == 3);
+    output.SetLayout(layout);
+    kernels::DynamicScratchpad scratchpad({}, AccessOrder(ws.stream()));
+    ctx_.gpu.stream = ws.stream();
```
[the remark this answers is not in the document]
Comment (stiepan) [reply dali/operators/image/color/equalize.cu]: done

**Pull request merged**